File: ios/Classes/DoordeckHeadlessSdkFlutterPlugin.swift
import Flutter
import UIKit
import DoordeckSDK

public class DoordeckHeadlessSdkFlutterPlugin: NSObject, FlutterPlugin {
  private var channel: FlutterMethodChannel!
  private let doordeckSdk: Doordeck

  override init() {
    self.doordeckSdk = KDoordeckFactory().initialize(sdkConfig: SdkConfig.Builder().build())
    super.init()
  }

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "doordeck_headless_sdk_flutter", binaryMessenger: registrar.messenger())
    let instance = DoordeckHeadlessSdkFlutterPlugin()
    instance.channel = channel
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "login":
      guard let args = call.arguments as? [String: Any],
            let email = args["email"] as? String,
            let password = args["password"] as? String else {
        result(FlutterError(code: "INVALID_ARGS", message: "Missing email or password", details: nil))
        return
      }
      login(email: email, password: password, result: result)

    case "setAuthToken":
      guard let args = call.arguments as? [String: Any],
            let token = args["authToken"] as? String else {
        result(FlutterError(code: "INVALID_ARGS", message: "Missing authToken", details: nil))
        return
      }
      setAuthToken(token: token, result: result)

    case "verify":
      guard let args = call.arguments as? [String: Any],
            let code = args["code"] as? String else {
        result(FlutterError(code: "INVALID_ARGS", message: "Missing code", details: nil))
        return
      }
      verify(code: code, result: result)

    case "logout":
      logout(result: result)

    case "getUserDetails":
      getUserDetails(result: result)

    case "getLocksBelongingToTile":
      guard let args = call.arguments as? [String: Any],
            let tileId = args["tileId"] as? String else {
        result(FlutterError(code: "INVALID_ARGS", message: "Missing tileId", details: nil))
        return
      }
      getLocksBelongingToTile(tileId: tileId, result: result)

    case "unlockDevice":
      guard let args = call.arguments as? [String: Any],
            let lockId = args["lockId"] as? String else {
        result(FlutterError(code: "INVALID_ARGS", message: "Missing lockId", details: nil))
        return
      }
      unlockDevice(lockId: lockId, result: result)

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func login(email: String, password: String, result: @escaping FlutterResult) {
    setKeyPairIfNeeded()
    doordeckSdk.accountless().login(email: email, password: password) { token, error in
      if let error = error {
        result(FlutterError(code: "LOGIN_ERROR", message: error.localizedDescription, details: error.localizedDescription))
      } else {
        self.respondNeedsVerification(result: result)
      }
    }
  }

  private func setAuthToken(token: String, result: @escaping FlutterResult) {
    setKeyPairIfNeeded()
    doordeckSdk.contextManager().setCloudAuthToken(token: token)
    respondNeedsVerification(result: result)
  }

  private func verify(code: String, result: @escaping FlutterResult) {
    doordeckSdk.account().verifyEphemeralKeyRegistration(code: code, privateKey: nil) { _, error in
      if let error = error {
        result(FlutterError(code: "VERIFY_ERROR", message: error.localizedDescription, details: error.localizedDescription))
      } else {
        result(nil)
      }
    }
  }

  private func logout(result: @escaping FlutterResult) {
    doordeckSdk.account().logout { error in
      if let error = error {
        result(FlutterError(code: "LOGOUT_ERROR", message: error.localizedDescription, details: error.localizedDescription))
      } else {
        result(nil)
      }
    }
  }

  private func getUserDetails(result: @escaping FlutterResult) {
    doordeckSdk.account().getUserDetails { response, error in
      if let error = error {
        result(FlutterError(code: "USER_DETAILS_ERROR", message: error.localizedDescription, details: error.localizedDescription))
      } else if let response = response {
        result(response.toNativeMap(
          userId: self.doordeckSdk.contextManager().getUserId(),
          certificateChainAboutToExpire: self.doordeckSdk.contextManager().isCertificateChainAboutToExpire(),
          tokenAboutToExpire: self.doordeckSdk.contextManager().isCloudAuthTokenAboutToExpire()
        ))
      } else {
        result(FlutterError(code: "USER_DETAILS_ERROR", message: "Unknown error", details: nil))
      }
    }
  }

  private func getLocksBelongingToTile(tileId: String, result: @escaping FlutterResult) {
    doordeckSdk.tiles().getLocksBelongingToTile(tileId: tileId) { response, error in
      if let error = error {
        result(FlutterError(code: "LOCKS_ERROR", message: error.localizedDescription, details: error.localizedDescription))
      } else if let response = response {
        result(response.toNativeMap())
      } else {
        result(FlutterError(code: "LOCKS_ERROR", message: "Unknown error", details: nil))
      }
    }
  }

  private func unlockDevice(lockId: String, result: @escaping FlutterResult) {
    let base = LockOperations.BaseOperation(
      userId: nil,
      userCertificateChain: nil,
      userPrivateKey: nil,
      lockId: lockId,
      notBefore: Int32(Date().timeIntervalSince1970),
      issuedAt: Int32(Date().timeIntervalSince1970),
      expiresAt: Int32(Date().addingTimeInterval(60).timeIntervalSince1970),
      jti: UUID().uuidString
    )
    let unlock = LockOperations.UnlockOperation(baseOperation: base, directAccessEndpoints: nil)

    doordeckSdk.lockOperations().unlock(unlockOperation: unlock) { error in
      if let error = error {
        result(FlutterError(code: "UNLOCK_ERROR", message: error.localizedDescription, details: nil))
      } else {
        result(nil)
      }
    }
  }

  private func setKeyPairIfNeeded() {
    if !doordeckSdk.contextManager().isKeyPairValid() {
      let kp = doordeckSdk.crypto().generateKeyPair()
      doordeckSdk.contextManager().setKeyPair(publicKey: kp.public_, privateKey: kp.private_)
    }
  }

  private func respondNeedsVerification(result: @escaping FlutterResult) {
    doordeckSdk.helper().assistedRegisterEphemeralKey(publicKey: nil) { response, error in
      if let error = error {
        result(FlutterError(code: "NEEDS_VERIFICATION_ERROR", message: error.localizedDescription, details: error.localizedDescription))
      } else if let response = response {
        result(response.toNativeMap())
      } else {
        result(FlutterError(code: "NEEDS_VERIFICATION_ERROR", message: "Unknown error", details: nil))
      }
    }
  }
}

// MARK: - Doordeck model extensions

extension AssistedRegisterEphemeralKeyResponse {
  func toNativeMap() -> NSDictionary {
    return ["requiresVerification": requiresVerification]
  }
}

extension TileLocksResponse {
  func toNativeMap() -> NSDictionary {
    return [
      "siteId": siteId,
      "tileId": tileId,
      "deviceIds": deviceIds
    ]
  }
}

extension UserDetailsResponse {
  func toNativeMap(userId: String?, certificateChainAboutToExpire: Bool, tokenAboutToExpire: Bool) -> NSDictionary {
    return [
      "userId": userId as Any,
      "certificateChainAboutToExpire": certificateChainAboutToExpire,
      "tokenAboutToExpire": tokenAboutToExpire,
      "publicKey": publicKey,
      "email": email,
      "displayName": displayName as Any,
      "emailVerified": emailVerified
    ]
  }
}

File: android/src/main/kotlin/com/doordeck/doordeckHeadlessSdkFlutter/DoordeckHeadlessSdkFlutterPlugin.kt
package com.doordeck.doordeckHeadlessSdkFlutter

import com.doordeck.multiplatform.sdk.ApplicationContext
import com.doordeck.multiplatform.sdk.Doordeck
import com.doordeck.multiplatform.sdk.KDoordeckFactory
import com.doordeck.multiplatform.sdk.config.SdkConfig
import com.doordeck.multiplatform.sdk.model.data.LockOperations
import com.doordeck.multiplatform.sdk.model.responses.AssistedRegisterEphemeralKeyResponse
import com.doordeck.multiplatform.sdk.model.responses.TileLocksResponse
import com.doordeck.multiplatform.sdk.model.responses.UserDetailsResponse
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result


/** DoordeckHeadlessSdkFlutterPlugin */
class DoordeckHeadlessSdkFlutterPlugin : FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel
    private lateinit var doordeckSdk: Doordeck

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "doordeck_headless_sdk_flutter")
        channel.setMethodCallHandler(this)

        doordeckSdk = KDoordeckFactory.initialize(
            SdkConfig.Builder()
                .setApplicationContext(ApplicationContext.apply { set(flutterPluginBinding.applicationContext)})
                .build()
        )
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        return when (call.method) {
            "login" -> login(call, result)
            "setAuthToken" -> setAuthToken(call, result)
            "verify" -> verify(call, result)
            "logout" -> logout(result)
            "getUserDetails" -> getUserDetails(result)
            "getLocksBelongingToTile" -> getLocksBelongingToTile(call, result)
            "unlockDevice" -> unlockDevice(call, result)
            else -> result.notImplemented()
        }

    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    /**
     * Auth operations
     */

    private fun login(call: MethodCall, result: Result) {
        val email = call.argument<String>("email")!!
        val password = call.argument<String>("password")!!
        setKeyPairIfNeeded()

        doordeckSdk.accountless().loginAsync(email, password)
            .thenApply {
                respondNeedsVerification(result)
            }
            .exceptionally { error ->
                result.error("LOGIN_ERROR", "", error)
            }
    }

    private fun setAuthToken(call: MethodCall, result: Result) {
        val authToken = call.argument<String>("authToken")!!
        setKeyPairIfNeeded()

        doordeckSdk.contextManager().setCloudAuthToken(authToken)
        respondNeedsVerification(result)
    }

    private fun verify(call: MethodCall, result: Result) {
        val code = call.argument<String>("code")!!
        doordeckSdk.account().verifyEphemeralKeyRegistrationAsync(code)
            .thenApply {
                result.success(null)
            }
            .exceptionally { error ->
                result.error("VERIFY_ERROR", "", error)
            }
    }

    private fun logout(result: Result) {
        doordeckSdk.account().logoutAsync()
            .thenApply {
                result.success(null)
            }
            .exceptionally { error ->
                result.error("LOGOUT_ERROR", "", error)
            }
    }

    /**
     * User operations
     */
    private fun getUserDetails(result: Result) {
        doordeckSdk.account().getUserDetailsAsync()
            .thenApply { response ->
                result.success(response.toNativeMap(
                    userId = doordeckSdk.contextManager().getUserId(),
                    certificateChainAboutToExpire = doordeckSdk.contextManager().isCertificateChainAboutToExpire(),
                    tokenAboutToExpire = doordeckSdk.contextManager().isCloudAuthTokenAboutToExpire(),
                ))
            }
            .exceptionally {  error ->
                result.error("USER_DETAILS_ERROR", "", error)
            }
    }

    /**
     * Device and tile operations
     */

    private fun getLocksBelongingToTile(call: MethodCall, result: Result) {
        val tileId = call.argument<String>("tileId")!!
        doordeckSdk.tiles().getLocksBelongingToTileAsync(tileId)
            .thenApply { response ->
                result.success(response.toNativeMap())
            }
            .exceptionally { error ->
                result.error("LOCKS_ERROR", "", error)
            }
    }

    private fun unlockDevice(call: MethodCall, result: Result) {
        val lockId = call.argument<String>("lockId")!!
        doordeckSdk.lockOperations().unlockAsync(
            LockOperations.UnlockOperation(
                LockOperations.BaseOperation(
                    lockId = lockId,
                )
            )
        )
            .thenApply {
                result.success(null)
            }
            .exceptionally { error ->
                result.error("UNLOCK_ERROR", "", error)
            }
    }

    private fun setKeyPairIfNeeded() {
        if (!doordeckSdk.contextManager().isKeyPairValid()) {
            val newKeyPair = doordeckSdk.crypto().generateKeyPair()
            doordeckSdk.contextManager().setKeyPair(
                publicKey = newKeyPair.public,
                privateKey = newKeyPair.private,
            )
        }
    }

    private fun respondNeedsVerification(result: Result) {
        doordeckSdk.helper().assistedRegisterEphemeralKeyAsync()
            .thenApply { response ->
                result.success(response.toNativeMap())
            }
            .exceptionally { error ->
                result.error("NEEDS_VERIFICATION_ERROR", "", error)
            }
    }

    private fun AssistedRegisterEphemeralKeyResponse.toNativeMap() = mapOf(
        "requiresVerification" to requiresVerification,
    )

    private fun TileLocksResponse.toNativeMap() = mapOf(
        "siteId" to siteId,
        "tileId" to tileId,
        "deviceIds" to deviceIds,
    )


    private fun UserDetailsResponse.toNativeMap(
        userId: String?,
        certificateChainAboutToExpire: Boolean,
        tokenAboutToExpire: Boolean,
    ) = mapOf(
        "userId" to userId,
        "certificateChainAboutToExpire" to certificateChainAboutToExpire,
        "tokenAboutToExpire" to tokenAboutToExpire,
        "publicKey" to publicKey,
        "email" to email,
        "displayName" to displayName,
        "emailVerified" to emailVerified,
    )

}
